### PR TITLE
nvprim op support runtime checks on dtype compatibility on prims.convert_element_type

### PR DESCRIPTION
**Pull request by jjsjann123**
I'm seeing issue that we lower `_to_copy` into `nvprims.convert_element_type`. In cases where we are casting to a dtype that's not supported by nvfuser, this raise runtime error.

I added a quick check in the lowering part where each op can peek at fx.node and make a runtime decision on whether the given op should be lowered to nvprim.

### CODE REVIEW

**Comment by jjsjann123** [comment]
Want to double check with @IvanYashchuk on his opinion with this PR.

**Review by IvanYashchuk — COMMENTED** [review]
I think `is_node_supported` is the right place to put the logic on whether a certain function is marked as supported by nvFuser or not.

But I think adding `nvprims.name.skip_nvfuser` is not ideal. The idea of "nvprims" namespace is that everything in it is "certified" to be supported by nvFuser execution. Of course, in the meantime, we have some gaps like in dtype conversion. What's your opinion on consolidating the logic inside the "is_node_supported" function, supposedly for this specific type conversion function it is a temporary solution.

In general, to skip or ignore any specified function, we could accept a list of functions to be ignored in "TorchRefsNvfuserCapabilityMode(ignore=list_or_dict)" and "NvfuserPrimsMode(ignore=list_or_dict)".

**Comment by jjsjann123** [comment]
> In general, to skip or ignore any specified function, we could accept a list of functions to be ignored in "TorchRefsNvfuserCapabilityMode(ignore=list_or_dict)" and "NvfuserPrimsMode(ignore=list_or_dict)".

Looks like we have this part implemented in #85811 I'm happy with that for debugging and WAR. Let's move forward with that one.

> But I think adding `nvprims.name.skip_nvfuser` is not ideal. The idea of "nvprims" namespace is that everything in it is "certified" to be supported by nvFuser execution. Of course, in the meantime, we have some gaps like in dtype conversion. What's your opinion on consolidating the logic inside the "is_node_supported" function, supposedly for this specific type conversion function it is a temporary solution.

Meanwhile, I do think it's handy for us to have a finer control on the lowering, just so we are more resistant to the support gap between prim vs nvfuser(nvprim).

> But I think adding `nvprims.name.skip_nvfuser` is not ideal.
^^^ Nailing down on this part, what's a better alternative in your opinion? I went with that so it's less invasive and we can have the skip function in the same file with the definition of nvfuser lowering. I'm very open to suggestions on the implementation.

:pray:

File: torch/_prims/nvfuser_executor.py [hunk 1]
```diff
@@ -208,6 +213,9 @@ def nvfuser_execute(gm: GraphModule, *args, executor_parameters=None):
 
 class NvfuserPrimOperatorSupport(torch.fx.passes.operator_support.OperatorSupport):
     def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
+        # special case to stop lowering to nvprim when converting to an unsupported type
+        if node.op == "call_function" and node.name == "convert_element_type":
```
Comment (IvanYashchuk) [inline torch/_prims/nvfuser_executor.py:217]: This wouldn't work if we had multiple dtype conversions in the graph.
`node.name` refers to the variable name, we need to compare `node.target == torch.ops.prims.convert_element_type.default`:
```py
In [1]: import torch

In [2]: a = torch.randn(3, 3)

In [3]: def func(a):
   ...:     return (a.to(torch.float16), a.to(torch.float64))
   ...:

In [4]: from torch._prims.context import TorchRefsMode

In [5]: from torch.fx.experimental.proxy_tensor import make_fx

In [6]: with TorchRefsMode():
   ...:     make_fx(func)(a).graph.print_tabular()
   ...:
opcode         name                    target                              args                                               kwargs
-------------  ----------------------  ----------------------------------  -------------------------------------------------  --------
placeholder    a_1                     a_1                                 ()                                                 {}
call_function  convert_element_type    prims.convert_element_type.default  (a_1, torch.float16)                               {}
call_function  convert_element_type_1  prims.convert_element_type.default  (a_1, torch.float64)                               {}
output         output                  output                              ((convert_element_type, convert_element_type_1),)  {}
```

Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py:217]: Oops. good catch! it's concerning that this is not caught by our test. I'll update it along with python tests.

Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py]: BTW, we are inserting in the segmentation here, which I feel is a little bit odd and we are going to be skipping nvprim.convert_element_type instead.

I think you are suggesting moving this to `_is_node_supported_nvfuser` inside `context.py` instead.

File: test/test_prims.py [hunk 2]
```diff
@@ -185,6 +185,38 @@ def func(a, b, c):
...
+        from torch._prims.executor import execute
+        from torch._prims.context import TorchRefsNvfuserCapabilityMode
+        from torch._prims_common import _torch_dtype_to_nvfuser_dtype_map
+
+        # initialize input as float32, which is different from `dtype` in the argument.
+        # this ensures that tracing will have a _to_copy node.
+        a = torch.randn(3, 3, device=device, dtype=torch.float32)
+
+        def func(x, dtype):
+            return x.to(dtype).to(x.dtype)
```
Comment (jjsjann123) [inline test/test_prims.py:201]: This has multiple cast to verify the previous bug. It also requires us to do an extra check on arg[0]...dtype

**Review by IvanYashchuk — APPROVED** [review]

**Comment by jjsjann123** [comment]
errr.. forgot that our stamps are just for emotional support.
Can I get some real developer to review and hopefully stamp this? cc'ing @ngimel @mruberry

**Review by ngimel — APPROVED** [review]
What types are unsupported by nvfuser? something like int16?
Does nvfuser support integer types for all ops? I remember opInfo/lowering tests were running only for fp16/fp32 inputs, everything else was offloaded to eager. 
Also what would happen with conversion to bfloat16 that's either supported or not depending on the hardware?

**Comment by jjsjann123** [comment]
> What types are unsupported by nvfuser? something like int16?

The one we run into are `uint8`, which @kevinstephano mentioned comes from the mask tensor. 

> Does nvfuser support integer types for all ops?

That was the goal, at least all ops that integer inputs makes sense. There might be missing support in runtime file, but we should generally just add support if they show up.

> I remember opInfo/lowering tests were running only for fp16/fp32 inputs, everything else was offloaded to eager. Also what would happen with conversion to bfloat16 that's either supported or not depending on the hardware?

bfloat16 is unfortuante. TS has some logic to block handling bf16 ops to nvfuser when device doesn't support. We don't have that for nvprim yet. cc'ing @IvanYashchuk Yet something more to plumb there.
